File: scripts/collect-hotel-photos.py
#!/usr/bin/env python3
"""
Hotel Photo Collection Script
Collects real hotel photos from multiple free sources
"""

import json
import requests
import time
import os
from typing import List, Dict
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class HotelPhotoCollector:
    def __init__(self):
        self.google_api_key = os.getenv('GOOGLE_PLACES_API_KEY')  # Set in environment
        self.collected_photos = {}
        
    def search_google_places_photos(self, hotel_name: str, city: str) -> List[str]:
        """Search for hotel photos using Google Places API"""
        if not self.google_api_key:
            logger.warning("Google Places API key not found")
            return []
            
        try:
            # Step 1: Find the place
            search_url = "https://maps.googleapis.com/maps/api/place/findplacefromtext/json"
            params = {
                'input': f"{hotel_name} {city} Nigeria",
                'inputtype': 'textquery',
                'fields': 'photos,place_id,name',
                'key': self.google_api_key
            }
            
            response = requests.get(search_url, params=params)
            data = response.json()
            
            if not data.get('candidates'):
                logger.info(f"No Google Places results for {hotel_name}")
                return []
                
            place = data['candidates'][0]
            photos = place.get('photos', [])
            
            if not photos:
                logger.info(f"No photos found for {hotel_name}")
                return []
            
            # Step 2: Get photo URLs
            photo_urls = []
            for photo in photos[:5]:  # Limit to 5 photos
                photo_url = f"https://maps.googleapis.com/maps/api/place/photo?maxwidth=800&photoreference={photo['photo_reference']}&key={self.google_api_key}"
                photo_urls.append(photo_url)
            
            logger.info(f"Found {len(photo_urls)} photos for {hotel_name}")
            return photo_urls
            
        except Exception as e:
            logger.error(f"Error searching Google Places for {hotel_name}: {e}")
            return []
    
    def search_website_photos(self, hotel_name: str, website_url: str) -> List[str]:
        """Scrape photos from hotel website"""
        if not website_url or website_url == 'nan':
            return []
            
        try:
            headers = {
                'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
            }
            response = requests.get(website_url, headers=headers, timeout=10)
            
            from bs4 import BeautifulSoup
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # Look for hotel images
            images = []
            
            # Common hotel image selectors
            selectors = [
                'img[alt*="hotel"]',
                'img[alt*="room"]',
                'img[src*="hotel"]',
                'img[src*="room"]',
                '.gallery img',
                '.hotel-images img',
                '.room-gallery img'
            ]
            
            for selector in selectors:
                img_tags = soup.select(selector)
                for img in img_tags:
                    src = img.get('src') or img.get('data-src')
                    if src:
                        # Convert relative URLs to absolute
                        if src.startswith('/'):
                            from urllib.parse import urljoin
                            src = urljoin(website_url, src)
                        images.append(src)
            
            # Remove duplicates and limit to 3 images
            unique_images = list(dict.fromkeys(images))[:3]
            logger.info(f"Found {len(unique_images)} images from {hotel_name} website")
            return unique_images
            
        except Exception as e:
            logger.error(f"Error scraping {hotel_name} website: {e}")
            return []
    
    def get_fallback_images(self, hotel_name: str, city: str, hotel_type: str) -> List[str]:
        """Get curated fallback images based on hotel characteristics"""
        
        # Curated hotel images by type and city
        fallback_images = {
            'luxury': [
                "https://images.unsplash.com/photo-1564501049412-61c2a3083791?w=800&h=600&fit=crop&auto=format&q=80",
                "https://images.unsplash.com/photo-1551882547-ff40c63fe5fa?w=800&h=600&fit=crop&auto=format&q=80",
                "https://images.unsplash.com/photo-1520250497591-112f2f40a3f4?w=800&h=600&fit=crop&auto=format&q=80"
            ],
            'business': [
                "https://images.unsplash.com/photo-1571003123894-1f0594d2b5d9?w=800&h=600&fit=crop&auto=format&q=80",
                "https://images.unsplash.com/photo-1582719478250-c89cae4dc85b?w=800&h=600&fit=crop&auto=format&q=80",
                "https://images.unsplash.com/photo-1519710164239-da123dc03ef4?w=800&h=600&fit=crop&auto=format&q=80"
            ],
            'apartment': [
                "https://images.unsplash.com/photo-1559599189-95f32f16b150?w=800&h=600&fit=crop&auto=format&q=80",
                "https://images.unsplash.com/photo-1562790351-d273a961e0e4?w=800&h=600&fit=crop&auto=format&q=80",
                "https://images.unsplash.com/photo-1512917774080-9991f1c4c750?w=800&h=600&fit=crop&auto=format&q=80"
            ]
        }
        
        # Determine category
        name_lower = hotel_name.lower()
        if any(word in name_lower for word in ['luxury', 'grand', 'royal', 'marriott', 'hilton']):
            category = 'luxury'
        elif hotel_type.lower() == 'apartment':
            category = 'apartment'
        else:
            category = 'business'
        
        return fallback_images[category]
    
    def collect_photos_for_hotel(self, hotel: Dict) -> Dict:
        """Collect photos for a single hotel"""
        hotel_name = hotel['name']
        city = hotel['city']
        website = hotel.get('website', '')
        hotel_type = hotel.get('type', 'Hotel')
        
        logger.info(f"Collecting photos for {hotel_name} in {city}")
        
        all_photos = []
        
        # 1. Try Google Places first (best quality)
        google_photos = self.search_google_places_photos(hotel_name, city)
        all_photos.extend(google_photos)
        
        # 2. Try hotel website if we need more photos
        if len(all_photos) < 3 and website:
            website_photos = self.search_website_photos(hotel_name, website)
            all_photos.extend(website_photos)
        
        # 3. Use curated fallbacks if still not enough
        if len(all_photos) < 5:
            fallback_photos = self.get_fallback_images(hotel_name, city, hotel_type)
            needed = 5 - len(all_photos)
            all_photos.extend(fallback_photos[:needed])
        
        # Update hotel data
        updated_hotel = hotel.copy()
        updated_hotel['images'] = all_photos[:5]  # Limit to 5 images
        
        # Rate limiting for API calls
        time.sleep(0.1)
        
        return updated_hotel
    
    def process_all_hotels(self, input_file: str, output_file: str):
        """Process all hotels and collect photos"""
        
        with open(input_file, 'r') as f:
            hotels = json.load(f)
        
        logger.info(f"Processing {len(hotels)} hotels...")
        
        updated_hotels = []
        
        for i, hotel in enumerate(hotels):
            try:
                updated_hotel = self.collect_photos_for_hotel(hotel)
                updated_hotels.append(updated_hotel)
                
                # Progress update
                if (i + 1) % 10 == 0:
                    logger.info(f"Processed {i + 1}/{len(hotels)} hotels")
                
            except Exception as e:
                logger.error(f"Error processing {hotel.get('name', 'Unknown')}: {e}")
                updated_hotels.append(hotel)  # Keep original if error
        
        # Save updated hotels
        with open(output_file, 'w') as f:
            json.dump(updated_hotels, f, indent=2)
        
        logger.info(f"Saved updated hotels to {output_file}")
        
        # Generate report
        self.generate_report(updated_hotels)
    
    def generate_report(self, hotels: List[Dict]):
        """Generate a report of photo collection results"""
        
        total_hotels = len(hotels)
        hotels_with_real_photos = 0
        hotels_with_fallback_only = 0
        
        for hotel in hotels:
            images = hotel.get('images', [])
            has_real_photos = any('googleapis.com' in img or hotel.get('website', '') in img for img in images)
            
            if has_real_photos:
                hotels_with_real_photos += 1
            else:
                hotels_with_fallback_only += 1
        
        report = f"""
HOTEL PHOTO COLLECTION REPORT
============================

Total Hotels: {total_hotels}
Hotels with Real Photos: {hotels_with_real_photos} ({hotels_with_real_photos/total_hotels*100:.1f}%)
Hotels with Fallback Only: {hotels_with_fallback_only} ({hotels_with_fallback_only/total_hotels*100:.1f}%)

To improve real photo coverage:
1. Set GOOGLE_PLACES_API_KEY environment variable
2. Contact hotels directly for photo permissions
3. Use hotel booking APIs (Booking.com, Expedia)
"""
        
        print(report)
        
        with open('photo-collection-report.txt', 'w') as f:
            f.write(report)

def main():
    collector = HotelPhotoCollector()
    
    # Process hotels
    collector.process_all_hotels(
        input_file='lib.hotels.json',
        output_file='lib.hotels.updated.json'
    )

if __name__ == "__main__":
    main()

File: scripts/collect-hotel-photos.js
#!/usr/bin/env node
/*
  Collect hotel photos using Google Places API.
  Usage:
    GOOGLE_PLACES_API_KEY=... node scripts/collect-hotel-photos.js --city=Owerri --budget=u80 --limit=20

  Output:
    scripts/output/hotel-photo-map.<city>.<budget>.json
*/

const fs = require('fs')
const path = require('path')

const API_KEY = process.env.GOOGLE_PLACES_API_KEY || process.env.NEXT_GOOGLE_PLACES_API_KEY
if (!API_KEY) {
  console.error('Missing GOOGLE_PLACES_API_KEY. Export it before running.')
  process.exit(1)
}

function parseArgs() {
  const args = process.argv.slice(2)
  const out = {}
  for (const a of args) {
    const m = a.match(/^--([^=]+)=(.*)$/)
    if (m) out[m[1]] = m[2]
  }
  return out
}

const args = parseArgs()
const CITY = args.city || 'ALL' // ALL = process all cities
const BUDGET = args.budget || '' // optional budget filter
const LIMIT = Number(args.limit || '50')

const ROOT = path.resolve(__dirname, '..')
const DATA_FILE = path.resolve(ROOT, 'lib.hotels.json')
const OUT_DIR = path.resolve(__dirname, 'output')
if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true })
const OUT_FILE = path.resolve(OUT_DIR, `hotel-photo-map.${CITY}.${BUDGET}.json`)

function sleep(ms){ return new Promise(r=>setTimeout(r, ms)) }

async function textSearch(q) {
  const url = new URL('https://maps.googleapis.com/maps/api/place/textsearch/json')
  url.searchParams.set('query', q)
  url.searchParams.set('key', API_KEY)
  const res = await fetch(url)
  if (!res.ok) throw new Error(`textsearch http ${res.status}`)
  return res.json()
}

function photoUrl(photo_reference, maxwidth=1200) {
  const u = new URL('https://maps.googleapis.com/maps/api/place/photo')
  u.searchParams.set('maxwidth', String(maxwidth))
  u.searchParams.set('photo_reference', photo_reference)
  u.searchParams.set('key', API_KEY)
  return u.toString()
}

async function main(){
  const raw = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'))
  const hotelsAll = raw.filter(h => h.type === 'Hotel' || h.type === 'Apartment')
  let hotels = CITY === 'ALL' ? hotelsAll : hotelsAll.filter(h => h.city === CITY)
  let filtered = hotels
  if (BUDGET === 'u80') {
    filtered = hotels.filter(h => Number(h.basePriceNGN || h.price || 0) < 80000)
  } else if (BUDGET === '80_130') {
    filtered = hotels.filter(h => Number(h.basePriceNGN || h.price || 0) >= 80000 && Number(h.basePriceNGN || h.price || 0) < 130000)
  } else if (BUDGET === '130_200') {
    filtered = hotels.filter(h => Number(h.basePriceNGN || h.price || 0) >= 130000 && Number(h.basePriceNGN || h.price || 0) < 200000)
  } else if (BUDGET === '200p') {
    filtered = hotels.filter(h => Number(h.basePriceNGN || h.price || 0) >= 200000)
  }
  filtered = filtered.slice(0, LIMIT)

  const map = {}
  for (const h of filtered) {
    const query = `${h.name}, ${h.city}, Nigeria`
    try {
      const data = await textSearch(query)
      const candidate = data.results && data.results[0]
      if (!candidate || !candidate.photos || candidate.photos.length === 0) {
        console.warn('No photos for', h.id, '-', h.name)
        continue
      }
      const picks = candidate.photos.slice(0, 4).map(p => photoUrl(p.photo_reference))
      map[h.id] = picks
      console.log('✓', h.id, picks.length, 'photos')
    } catch (e) {
      console.warn('ERR', h.id, e.message)
    }
    // Be polite with API
    await sleep(250)
  }

  fs.writeFileSync(OUT_FILE, JSON.stringify(map, null, 2))
  console.log('\nSaved mapping to', OUT_FILE)
}

main().catch(err => {
  console.error(err)
  process.exit(1)
})
